File: unisync_backend/lib/core/device_provider.dart
// ignore_for_file: close_sinks

import 'dart:io';

import 'package:unisync_backend/core/device.dart';
import 'package:unisync_backend/core/device_connection.dart';
import 'package:unisync_backend/notification/event_notifier.dart';
import 'package:unisync_backend/notification/events/device_connected.event.dart';
import 'package:unisync_backend/utils/logger.dart';

import '../models/device_info/device_info.model.dart';
import '../notification/events/device_disconnected.event.dart';

class DeviceProvider {
  DeviceProvider._();

  static final _devices = <DeviceInfo, Device>{};

  static List<DeviceInfo> get devices => _devices.values.map((e) => e.info).toList();

  static void create({required DeviceInfo info, required SecureSocket socket}) {
    if (_devices.containsKey(info)) {
      socket.close();
    } else {
      _devices[info] = Device(DeviceConnection(socket, info));
      infoLog('Connected to ${info.name} (${info.ip}).');
      UnisyncEventNotifier.publish(DeviceConnectedEvent(info));
    }
  }

  static void remove(DeviceInfo info) {
    _devices.remove(info.id);
    UnisyncEventNotifier.publish(DeviceDisconnectedEvent(info));
  }
}

File: unisync_android/app/src/main/java/com/anhquan/unisync/core/providers/DeviceProvider.kt
package com.anhquan.unisync.core.providers

import android.content.Context
import com.anhquan.unisync.core.device.Device
import com.anhquan.unisync.core.device.dependencies.DeviceConnection
import com.anhquan.unisync.core.device.dependencies.PairingHandler
import com.anhquan.unisync.models.DeviceInfo
import com.anhquan.unisync.utils.infoLog
import io.reactivex.rxjava3.subjects.PublishSubject
import javax.net.ssl.SSLSocket

object DeviceProvider {
    private val _devices = mutableMapOf<DeviceInfo, Device>()
    val devices: List<DeviceInfo> get() = _devices.keys.toList()

    data class DeviceNotification(
        val device: DeviceInfo,
        val connected: Boolean = true,
        val pairState: PairingHandler.PairState? = null
    )

    val deviceNotifier = PublishSubject.create<DeviceNotification>()

    fun create(
        context: Context,
        info: DeviceInfo,
        socket: SSLSocket
    ) {
        if (!_devices.containsKey(info)) {
            _devices[info] = Device(context, DeviceConnection(socket), info)
            infoLog("${this::class.simpleName}: Connected to ${info.name}@${info.ip}.")
        } else {
            socket.close()
            infoLog("${this::class.simpleName}: Duplicate connection to ${info.name}@${info.ip}. Closing connection...")
        }
    }

    fun remove(info: DeviceInfo) {
        _devices.remove(info)
        deviceNotifier.onNext(
            DeviceNotification(info, false)
        )
    }

    fun get(info: DeviceInfo): Device? {
        return _devices[info]
    }

    fun get(deviceId: String): Device? {
        val key = _devices.keys.firstOrNull {
            it.id == deviceId
        }
        return if (key == null) null else _devices[key]
    }
}